pkg/storage/unified/resource/search_server_distributor.go

```diff
@@ -114,6 +119,94 @@ func (ds *distributorServer) GetStats(ctx context.Context, r *resourcepb.Resourc
 	return client.GetStats(ctx, r)
 }
 
+func (ds *distributorServer) RebuildIndexes(ctx context.Context, r *resourcepb.RebuildIndexesRequest) (*resourcepb.RebuildIndexesResponse, error) {
+	ctx, span := ds.tracing.Start(ctx, "distributor.RebuildIndexes")
+	defer span.End()
+
+	// distribute the request to all search pods to minimize risk of stale index
+	// it will not rebuild on those which don't have the index open
+	rs, err := ds.ring.GetAllHealthy(searchRingRead)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get all healthy instances from the ring")
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		md = make(metadata.MD)
+	}
+
+	var wg sync.WaitGroup
+	var detailsMu sync.Mutex
+	var details strings.Builder
+	var totalRebuildCount atomic.Int64
+	errorCh := make(chan error, len(rs.Instances))
+
+	for _, inst := range rs.Instances {
+		wg.Add(1)
+		go func(inst ring.InstanceDesc) {
+			defer wg.Done()
+
+			client, err := ds.clientPool.GetClientForInstance(inst)
+			if err != nil {
+				ds.log.Error("failed to get client for instance", "instance", inst.Id, "err", err)
+				errorCh <- fmt.Errorf("instance %s: failed to get client, %w", inst.Id, err)
+				return
+			}
+
+			// context for each req
+			instanceCtx := userutils.InjectOrgID(
+				metadata.NewOutgoingContext(ctx, md),
+				r.Namespace,
+			)
+
+			err = grpc.SetHeader(instanceCtx, metadata.Pairs("proxied-instance-id", inst.Id))
+			if err != nil {
+				ds.log.Debug("error setting grpc header", "err", err)
+			}
+
+			rsp, err := client.(*RingClient).Client.RebuildIndexes(instanceCtx, r)
+			if err != nil {
+				ds.log.Error("failed to distribute rebuild index request", "instance", inst.Id, "error", err)
+				errorCh <- fmt.Errorf("instance %s: failed to distribute rebuild index request, %w", inst.Id, err)
+				return
+			}
+
+			if rsp.Error != nil {
+				errorCh <- fmt.Errorf("instance %s: rebuild index request returned the error %s", inst.Id, rsp.Error.Message)
+				return
+			}
+
+			if rsp.Details != "" {
+				detailsMu.Lock()
```

Comment:
To simplify concurrency, we could pass *all* responses to the channel, and build details/rebuild count in the outer goroutine, similar how we pass errors.